src/middlewared/middlewared/plugins/vm.py

```diff
@@ -7,481 +6,609 @@
 
-    async def status(self, id):
-        supervisor = self._vm.get(id)
-        if supervisor and await supervisor.running():
-            return {
-                'state': 'RUNNING',
-                'pid': supervisor.proc.pid if supervisor.proc else None,
-            }
+        if not self.connection or not self.connection.isAlive():
+            raise CallError(f'Failed to connect to libvirtd for {self.vm_data["name"]}')
+
+        self.domain = None
+        self.update_domain()
+
+    def update_domain(self, vm_data=None):
+        # This can be called to update domain to reflect any changes introduced to the VM
+        self.vm_data = vm_data or self.vm_data
+        try:
+            self.domain = self.connection.lookupByName(self.libvirt_domain_name)
+        except libvirt.libvirtError:
+            self.domain = None
         else:
-            return {
-                'state': 'STOPPED',
-                'pid': None,
-            }
+            if not self.domain.isActive():
+                # We have a domain defined and it is not running
+                self.undefine_domain()
+
+        if not self.domain:
+            self.__define_domain()
+
+    def status(self):
+        domain = self.domain
+        return {
+            'state': 'STOPPED' if not domain.isActive() else 'RUNNING',
+            'pid': None if not domain.isActive() else self.domain.ID(),
+            'domain_state': DomainState(domain.state()[0]).name,
+        }
+
+    def __define_domain(self):
+        if self.domain:
+            raise CallError(f'{self.libvirt_domain_name} domain has already been defined')
 
+        vm_xml = etree.tostring(self.construct_xml()).decode()
+        if not self.connection.defineXML(vm_xml):
+            raise CallError(f'Unable to define persistent domain for {self.libvirt_domain_name}')
 
-class VMSupervisor(object):
+        self.domain = self.connection.lookupByName(self.libvirt_domain_name)
 
-    def __init__(self, manager, vm):
-        self.manager = manager
+    def undefine_domain(self):
+        if self.domain.isActive():
+            raise CallError(f'Domain {self.libvirt_domain_name} is active. Please stop it first')
 
-        os.makedirs('/var/log/vm', exist_ok=True)
-        self.logger = self.manager.logger.getChild(f'vm_{vm["id"]}')
+        if self.vm_data['bootloader'] == 'GRUB':
+            shutil.rmtree(
+                os.path.join('/tmp/grub', self.libvirt_domain_name), ignore_errors=True
+            )
+
+        self.domain.undefine()
+        self.domain = None
+
+    def __getattribute__(self, item):
+        retrieved_item = object.__getattribute__(self, item)
+        if callable(retrieved_item) and item in ('start', 'stop', 'restart', 'poweroff', 'undefine_domain', 'status'):
+            if not getattr(self, 'domain', None):
+                raise CallError('Domain attribute not defined, please re-instantiate the VM class')
```

Comment:
I doubt that this internal logic errors should be `CallError`. After a while, we would be forced to i18n all this, and translating programming errors would have no sense. Can it be a `RuntimeError` for example since user is not responsible for this and can't fix anything?